libyara/exception.h

```diff
@@ -142,7 +142,49 @@ static void exception_handler(int sig, siginfo_t * info, void *context)
         (jmp_buf*) yr_thread_storage_get_value(&yr_trycatch_trampoline_tls);
 
     if (jb_ptr != NULL)
+    {
       siglongjmp(*jb_ptr, 1);
+      // The long jump means the following code to invoke the old exception handler is never executed
+    }
+
+    // If we're here, the signal we received didn't originate from YARA.
+    // In this case, we want to invoke the original signal handler, which may handle the signal.
+
+    // Lock the exception handler mutex to prevent simultaneous write access while we read the old signal handler
+    pthread_mutex_lock(&exception_handler_mutex);
+    struct sigaction old_handler;
+    if (sig == SIGBUS)
+      old_handler = old_sigbus_exception_handler;
+    else
+      old_handler = old_sigsegv_exception_handler;
+    pthread_mutex_unlock(&exception_handler_mutex);
+
+    if (old_handler.sa_flags & SA_SIGINFO)
+    {
+      old_handler.sa_sigaction(sig, info, context);
+    }
+    else
+    {
+      if (old_handler.sa_handler == SIG_DFL)
+      {
+        // Old handler is the default action. To do this, set the signal handler back to default and raise the signal.
```

Comment:
We **might** get away without locking and having to worry about non-atomic operations if we use `SA_RESETHAND` when installing our handler over SIG_DFL handlers.

Reply:
`SA_RESETHAND` comes with additional complications, though. If we use it, we are inherently racy when two threads run (within YARA) into a `SIGBUS` in parallel: the first occurrence will reset the signal handler and the second one will therefore no longer be caught by our signal handler, and instead cause a crash.